src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs

```diff
@@ -54,99 +53,92 @@ protected JSRuntime()
+    internal async ValueTask<TValue> InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(long targetInstanceId, string identifier, JsonSerializerOptions? options, object?[]? args)
     {
         if (DefaultAsyncTimeout.HasValue)
         {
             using var cts = new CancellationTokenSource(DefaultAsyncTimeout.Value);
             // We need to await here due to the using
-            return await InvokeAsync<TValue>(targetInstanceId, identifier, cts.Token, args);
+            return await InvokeAsync<TValue>(targetInstanceId, identifier, options, cts.Token, args);
         }
 
-        return await InvokeAsync<TValue>(targetInstanceId, identifier, CancellationToken.None, args);
+        return await InvokeAsync<TValue>(targetInstanceId, identifier, options, CancellationToken.None, args);
     }
 
+    internal ValueTask<TValue> InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(long targetInstanceId, string identifier, object?[]? args)
+        => InvokeAsync<TValue>(targetInstanceId, identifier, options: null, args);
+
     [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "We expect application code is configured to ensure JS interop arguments are linker friendly.")]
     internal ValueTask<TValue> InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(
         long targetInstanceId,
         string identifier,
+        JsonSerializerOptions? jsonSerializerOptions,
         CancellationToken cancellationToken,
         object?[]? args)
     {
-        var taskId = Interlocked.Increment(ref _nextPendingTaskId);
-        var tcs = new TaskCompletionSource<TValue>();
-        if (cancellationToken.CanBeCanceled)
+        if (cancellationToken.IsCancellationRequested)
         {
-            _cancellationRegistrations[taskId] = cancellationToken.Register(() =>
-            {
-                tcs.TrySetCanceled(cancellationToken);
-                CleanupTasksAndRegistrations(taskId);
-            });
+            return ValueTask.FromCanceled<TValue>(cancellationToken);
         }
-        _pendingTasks[taskId] = tcs;
 
-        try
+        var taskId = Interlocked.Increment(ref _nextPendingTaskId);
+        var interopTask = new JSInteropTask<TValue>(cancellationToken, onCanceled: () =>
         {
-            if (cancellationToken.IsCancellationRequested)
-            {
-                tcs.TrySetCanceled(cancellationToken);
-                CleanupTasksAndRegistrations(taskId);
+            _pendingTasks.TryRemove(taskId, out _);
+        });
 
-                return new ValueTask<TValue>(tcs.Task);
-            }
+        _pendingTasks[taskId] = interopTask;
 
-            var argsJson = args is not null && args.Length != 0 ?
-                JsonSerializer.Serialize(args, JsonSerializerOptions) :
-                null;
+        try
+        {
             var resultType = JSCallResultTypeHelper.FromGeneric<TValue>();
+            jsonSerializerOptions ??= JsonSerializerOptions;
```

Comment:
I saw you were combining the `IJsonTypeInfoResolver` from the passed in options with the ones from `JsonSerializerOptions` in previous iterations. Why did you decide that's no longer necessary? Is the idea that no one who passes custom options is going to need to pass a `DotNetStreamReference` argument or anything like that? Or is it easy to make stuff like that work without us combining it for you?

Reply:
I removed that because creating a combined `IJsonTypeInfoResolver` here means we'd get a new `JsonSerializerOptions` that we need to cache and manage the lifetime of. We decided it was better to have the developer pass in their own `JsonSerializerOptions` so we can avoid that problem.

You can still add something like [this](https://github.com/dotnet/aspnetcore/blob/04910bbbb15f78e245be8bfdec6a3edea32518fd/src/Components/Shared/src/JsonSerialization/JsonConverterFactoryTypeInfoResolver.cs) to the `TypeInfoResolverChain` where `T` is the type that you want to be resolved using a custom converter. Alternatively, you can add the type directly to the `JsonSerializerContext`, but that generates unnecessary code to handle structural serialization, which is often not what you want (especially in cases like `DotNetObjectReference<MyMassiveObjectThatIDoNotWantToBeSerialized>`).

Reply:
`DotNetStreamReferenceJsonConverter` is internal, so it's not clear how you could add it to your own `JsonSerializerOptions` unless you could copy it from another options instance. I guess you could try copying it from `JSRuntime.JsonSerializerOptions` since it's a `protected internal` property on a public type which makes it [public API](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/protected-internal), but that seems really hacky.

Wouldn't the easiest thing be to add all the `JSRuntime` converters except the `ByteArrayJsonConverter` (which looks super breaking to use for anything other than `JSRuntime`) to `JsonOptions.JsonSerializerOptions` and then use that? If we're afraid that `JsonOptions.JsonSerializerOptions` is too generic, we could add a `JSRuntimeJsonOptions` type and use that instead. Then we could still include `ByteArrayJsonConverter` by default.

Reply:
> it's not clear how you could add it to your own JsonSerializerOptions unless you could copy it from another options instance

I thought that's what you *must* do, i.e., clone and extend the options from whatever `IJSRuntime` you're going to call. That's much safer than having a globally-shared set of options and hoping the ecosystem will agree on all the details.

My only concern is whether we can detect cases where the developer passes a `JsonSerializerOptions` that wasn't copied from an `IJSRuntime`, since it would be confusing for features like `ElementReference` not to work.

Reply:
If it's possible for each `JSRuntime` type to put some kind of marker into the options that lets it quickly detect whether any other `JsonSerializerOptions` was cloned from its defaults, that would be ideal.